drivers/adc/adc_lmp90xxx.c

```diff
@@ -0,0 +1,1136 @@
+#define LMP90XXX_REG_ADC_DONE        0x18U
+#define LMP90XXX_REG_SENDIAG_FLAGS   0x19U
+#define LMP90XXX_REG_ADC_DOUT        0x1AU
+#define LMP90XXX_REG_SPI_CRC_DAT     0x1DU
+#define LMP90XXX_REG_CH_STS          0x1EU
+#define LMP90XXX_REG_CH_SCAN         0x1FU
+
+/* LMP90xxx channel input and configuration registers */
+#define LMP90XXX_REG_CH_INPUTCN(ch) (0x20U + (2 * ch))
+#define LMP90XXX_REG_CH_CONFIG(ch)  (0x21U + (2 * ch))
+
+/* LMP90xxx upper (URA) and lower (LRA) register addresses */
+#define LMP90XXX_URA(addr) ((addr >> 4U) & GENMASK(2, 0))
+#define LMP90XXX_LRA(addr) (addr & GENMASK(3, 0))
+
+/* LMP90xxx instruction byte 1 (INST1) */
+#define LMP90XXX_INST1_WAB 0x10U
+#define LMP90XXX_INST1_RA  0x90U
+
+/* LMP90xxx instruction byte 2 (INST2) */
+#define LMP90XXX_INST2_WB        0U
+#define LMP90XXX_INST2_R         BIT(7)
+#define LMP90XXX_INST2_SZ_1      (0x0U << 5)
+#define LMP90XXX_INST2_SZ_2      (0x1U << 5)
+#define LMP90XXX_INST2_SZ_3      (0x2U << 5)
+#define LMP90XXX_INST2_SZ_STREAM (0x3U << 5)
+
+/* LMP90xxx register values/commands */
+#define LMP90XXX_REG_AND_CNV_RST     0xC3U
+#define LMP90XXX_SDO_DRDYB_DRIVER(x) ((x & BIT_MASK(3)) << 1)
+#define LMP90XXX_PWRCN(x)            (x & BIT_MASK(2))
+#define LMP90XXX_RTD_CUR_SEL(x)      (x & BIT_MASK(4))
+#define LMP90XXX_SPI_DRDYB_D6(x)     ((x & BIT(0)) << 7)
+#define LMP90XXX_EN_CRC(x)           ((x & BIT(0)) << 4)
+#define LMP90XXX_DRDYB_AFT_CRC(x)    ((x & BIT(0)) << 2)
+#define LMP90XXX_CH_SCAN_SEL(x)      ((x & BIT_MASK(2)) << 6)
+#define LMP90XXX_LAST_CH(x)          ((x & BIT_MASK(3)) << 3)
+#define LMP90XXX_FIRST_CH(x)         (x & BIT_MASK(3))
+#define LMP90XXX_BURNOUT_EN(x)       ((x & BIT(0)) << 7)
+#define LMP90XXX_VREF_SEL(x)         ((x & BIT(0)) << 6)
+#define LMP90XXX_VINP(x)             ((x & BIT_MASK(3)) << 3)
+#define LMP90XXX_VINN(x)             (x & BIT_MASK(3))
+#define LMP90XXX_BGCALN(x)           (x & BIT_MASK(3))
+#define LMP90XXX_ODR_SEL(x)          ((x & BIT_MASK(3)) << 4)
+#define LMP90XXX_GAIN_SEL(x)         ((x & BIT_MASK(3)) << 1)
+#define LMP90XXX_BUF_EN(x)           (x & BIT(0))
+
+/* Invalid (never used) Upper Register Address */
+#define LMP90XXX_INVALID_URA UINT8_MAX
+
+/* Maximum number of ADC channels */
+#define LMP90XXX_MAX_CHANNELS 7
+
+/* Maximum number of ADC inputs */
+#define LMP90XXX_MAX_INPUTS 8
+
+/* Default Output Data Rate (ODR) is 214.65 SPS */
+#define LMP90XXX_DEFAULT_ODR 7
+
+/* Macro for checking if Data Ready Bar IRQ is in use */
+#define LMP90XXX_HAS_DRDYB(config) (config->drdyb_dev_name != NULL)
+
+struct lmp90xxx_config {
+	const char *spi_dev_name;
+	const char *spi_cs_dev_name;
+	u8_t spi_cs_pin;
+	const char *drdyb_dev_name;
+	u32_t drdyb_pin;
+	int drdyb_flags;
+	u8_t rtd_current;
+	u32_t spi_freq;
+	u16_t spi_slave;
+	u8_t resolution;
+	u8_t channels;
+};
+
+struct lmp90xxx_data {
+	struct device *dev;
+	struct adc_context ctx;
+	struct device *spi_dev;
+	struct spi_cs_control spi_cs;
+	struct spi_config spi_cfg;
```

Comment:
This structure is filled only with constant values known at the build time, so it could be moved to `lmp90xxx_config`.

Reply:
Unfortunately there is one field in `struct spi_config` that is set at run-time (the `cs` field, which contains a pointer to the CS GPIO `struct device`).

Reply:
But the `cs` field can also be filled at build-time, to `NULL` or to a `struct spi_cs_control` pointer, basing on the `DT_INST_##n##_TI_LMP##t##_CS_GPIOS_CONTROLLER` being `NULL` or not.

Reply:
@anangl Ah, got it. I have moved the `struct spi_config` instance to `struct lmp90xxx_config` as suggested.